File: apps/silver_brain_core/lib/item_store.ex
defprotocol SilverBrain.Core.ItemStore do
  import SilverBrain.Core

  alias SilverBrain.Core.Item

  @type result(t) :: SilverBrain.Core.result(t)
  @type result :: SilverBrain.Core.result()

  # ============================================================
  #  Item
  # ============================================================

  defmodule UpdateItem do
    use TypedStruct

    typedstruct do
      field :id, String.t()
      field :name, String.t()
    end
  end

  @spec get_item(t, Item.id()) :: result(Item.t())
  def get_item(store, item_id)

  @spec get_item(t, Item.id(), [atom()]) :: result(Item.t())
  def get_item(store, item_id, select)

  @spec get_items(t, list(Item.id()), select: atom()) :: result([Item.t()])
  def get_items(store, item_ids, select)

  @spec create_item(t, String.t()) :: result(Item.id())
  def create_item(store, name)

  @spec update_item(t, UpdateItem.t()) :: result()
  def update_item(store, item)

  @spec delete_item(t, Item.id()) :: result()
  def delete_item(store, item_id)

  # ============================================================
  #  Property
  # ============================================================

  @spec create_item_property(t, Item.id(), String.t(), String.t()) :: result()
  def create_item_property(store, item_id, key, value)

  @spec delete_item_property(t, Item.id(), String.t()) :: result()
  def delete_item_property(store, item_id, key)

  # ============================================================
  #  Link
  # ============================================================

  @spec create_child(t, Item.id(), Item.id()) :: result()
  def create_child(store, parent_id, child_id)

  @spec get_parents(t, Item.id()) :: result([Item.id()])
  def get_parents(store, item_id)

  @spec get_children(t, Item.id()) :: result([Item.id()])
  def get_children(store, item_id)

  @spec delete_child(t, Item.id(), Item.id()) :: result()
  def delete_child(store, parent_id, child_id)
end

File: modules/store/src/main/scala/item_store.scala
package silver_brain.store

import silver_brain.core.*
import silver_brain.store.repo.*

import cats.*
import cats.effect.*
import cats.implicits.*
import doobie.*
import doobie.implicits.*
import doobie.util.transactor.Transactor
import cats.effect.IO
import com.github.ksuid.Ksuid
import java.time.Instant

class SqlItemStore(
    private val transactor: Transactor[IO],
    private val storeName: String
) extends ItemStore:

  // ============================================================
  //  Item
  // ============================================================

  def getItem(
      itemId: String,
      loadOptions: ItemLoadOptions
  ): IO[StoreResult[Item]] =
    for rowOpt <- ItemRepo
        .getOne(itemId)
        .option
        .transact(this.transactor)
    yield rowOpt match
      case Some(row) => Right(row.toItem(loadOptions))
      case None      => Left(StoreError.IdNotFound(itemId))

  def getItems(
      itemIds: Seq[String],
      loadOptions: ItemLoadOptions
  ): IO[StoreResult[Seq[Item]]] =
    if itemIds.isEmpty then
      IO.pure(Left(StoreError.InvalidArgument("Empty id list")))
    else
      for result <- ItemRepo.getMany(itemIds).to[List].transact(this.transactor)
      yield Right(result.map(_.toItem(loadOptions)))

  def searchItems(
      search: String,
      loadOptions: ItemLoadOptions
  ): StoreResult[Seq[Item]] = ???

  def createItem(item: CreateItemArgs): IO[StoreResult[String]] =
    val id = "i_" + Ksuid.newKsuid().toString()

    for itemId <- ItemRepo
        .create(id, item, Instant.now())
        .run
        .transact(this.transactor)
    yield Right(id)

  def updateItem(item: UpdateItemArgs): IO[StoreResult[Unit]] =
    for updatedCount <- ItemRepo
        .update(item, Instant.now())
        .run
        .transact(this.transactor)
    yield
      if updatedCount == 0 then Left(StoreError.IdNotFound(item.id))
      else Right(())

  def deleteItem(itemId: String): IO[StoreResult[Unit]] =
    for _ <- ItemRepo.delete(itemId).run.transact(this.transactor)
    yield Right(())

  // ============================================================
  //  Property
  // ============================================================

  def upsertItemProperty(
      itemId: String,
      key: String,
      value: String
  ): StoreResult[Unit] = ???

  def deleteItemProperty(itemId: String, key: String): StoreResult[Unit] = ???

  // ============================================================
  //  Link
  // ============================================================

  def createLink(parent: String, child: String): StoreResult[Unit] = ???

  def getParents(itemId: String): StoreResult[Seq[String]] = ???

  def getChildren(itemId: String): StoreResult[Seq[String]] = ???

  def deleteLink(parent: String, child: String): StoreResult[Unit] = ???

  // ============================================================
  //  Reference
  // ============================================================

  def createReference(
      source: String,
      target: String,
      annotation: String
  ): StoreResult[String] = ???

  def getReference(referenceId: String): StoreResult[Reference] = ???

  def getReferences(referenceIds: Seq[String]): StoreResult[Seq[Reference]] =
    ???

  def updateReference(
      referenceId: String,
      annotation: String
  ): StoreResult[Unit] = ???

  def deleteReference(referenceId: String): StoreResult[Unit] = ???
